rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/parametric/model/ParametricFileFacts.java

```diff
@@ -108,66 +134,228 @@ public void close(ISourceLocation loc) {
+        }
+
+        public void calculateAnalyzer(int version, Duration delay) {
+            calculateAnalyzerLatestResult = calculate(version, latestVersionCalculateAnalyzer, delay, () -> {
+                var analysis = summaryAnalyzer.calculateSummary();
+                var analysisTree = analysis.input;
+                var analysisMessages = analysis.output.get().thenApply(Supplier::get);
+                return analysisTree.thenCombine(analysisMessages, (tree, messages) -> {
+                    if (analysis.calculation.isInterrupted()) {
+                        return Optional.<VersionedDiagnostics>empty();
+                    } else {
+                        return Optional.of(reportDiagnostics(diagnosticsAnalyzer, tree, messages));
+                    }
+                });
+            });
+        }
+
+        /**
+         * The main challenge when running the builder is that it might produce
+         * a subset of the same diagnostics as the analyzer. Thus, to avoid
+         * showing the same diagnostics twice, a diff needs to be computed of
+         * the latest analyzer diagnostics and the current builder diagnostics.
+         * To do this in a reliable way, each builder calculation awaits the
+         * completion of the latest analyzer calculation, such that:
+         *   - the builder can use exactly the same AST as the analyzer;
+         *   - the builder has access to the diagnostics of the analyzer.
+         */
+        public void calculateBuilder(int version, Duration delay) {
+
+            // If an analyzer calculation hasn't been scheduled yet (ever), then
+            // it needs to be scheduled first.
+            if (calculateAnalyzerLatestResult == null) {
+                calculateAnalyzer(version, Duration.ZERO);
+                calculateBuilder(version, delay);
+            }
+
+            // If an analyzer calculation has been scheduled before, then the
+            // builder can await its completion.
+            else {
+                calculateAnalyzerLatestResult.thenAccept(result -> {
+
+                    // If the analyzer calculation wasn't abandoned (due to
+                    // debouncing) or interrupted, then diagnostics of the
+                    // analyzer are available, so proceed.
+                    if (result.isPresent()) {
+                        var analysis = result.get();
+                        calculate(version, latestVersionCalculateBuilder, delay, () -> {
+
+                            // Use exactly the same AST as in the analyzer
+                            // calculation. In this way, a reliable diff of the
+                            // latest analyzer diagnostics and the current
+                            // builder diagnostics can be computed (by removing
+                            // the former from the latter).
+                            var build = summaryBuilder.calculateSummary(CompletableFuture.completedFuture(analysis.tree));
```

Comment:
what if a user saves a file 2 in succession, in between the saves the analyzer was finished, but the builder was not started yet, wouldn't that still cause a race? or are the previous saves always interrrupted?

(since it appears that it's depending on the success of the analyze, not on that of the builder)